chevah/chevah-keycert
# Add support for ECDSA and ED

Scope
=====

Fixes #42
Fixes #39
Fixes #38

This adds support for ECDSA and ED private keys.

It support OpenSSH and PuttyFormat.

Old OpenSSH format and SSH.COM format only support RSA and DSA.

This also removes the usage of pycrypto.

Changes
=======

Use the code from upstream Twisted, revision 38a1ed6f9 

As a side effect, this stops using pycrypto and uses crypography instead. 

I have updated  the keycert-demo.py command to allow re-encoding a file after loading and encrypting the re-encoded file.
The demo also has support for signing and validating the signature.


How to try and test the changes
===============================

reviewers: @dumol @lgheorghiu 

There is a lot to test here :)

Version 2.0.0 was released on our private PyPI to help with chevah/server testing and Laura will test the server integration.


1.Load EC and ED keys
------------

Try to generate ECDSA and ED keys using OpenSSH and PuttyGen and check that you can load them... with and without password. This is for public and private keys.

There is also support for loading ECDSA and ED keys from OpenSSL X509 PEM public keys.

Then there is also the `PKCS#8` (public and private) and public `PKCS#1` :)

```
# Load ECDSA and output as OpenSSH without password
./build-keycert/bin/python keycert-demo.py ssh-load-key --file ecdsa-521.ppk

# Load password protected ECDSA key and output as PuTTY format without password 
./build-keycert/bin/python keycert-demo.py ssh-load-key --file ecdsa-521.ppk --type putty --password chevah
```

2.Generate EC and ED keys
----------------------------------

Check key generation help CLI as this also goes to chevah/server

```
$ ./build-keycert/bin/python keycert-demo.py ssh-gen-key --help
```

Generate the key in OpenSSH format and then check that you can load them / convert them with OpenSSH or Putty.

```
$ ./build-keycert/bin/python keycert-demo.py ssh-gen-key --key-file id_ecdsa --key-type ecdsa --key-size=384
```

Generate  key in putty format with password protection

```
$ ./build-keycert/bin/python keycert-demo.py ssh-gen-key --key-type ed25519 --key-format putty --key-password chevah
```


3.Sign and verify data
--------------

```
# Generate the signature in base 64 using private key.
$ ./build-keycert/bin/python keycert-demo.py ssh-sign-data --file id_ecdsa --data test-value

# Validate the signature against the original data and public key.
$ ./build-keycert/bin/python keycert-demo.py ssh-verify-data --file id_ecdsa.pub --data test-value --signature AAAAE2VjZHNhLXNoYTItbmlzdHAzODQAAABpAAAAMQC2h6Akl/26ln+WxWqW8u38YQlUaR9rMt24E6/9Cl751UNQ0/e3xKzPMp75A6GZnIMAAAAwCCikhECXzFc2j6/+4nJ0VzBUJDi2WRS+OJiVozzlrNCMf3er2lM3lp4QLPP1ltLj
